inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/util/QueryUtil.java

```diff
@@ -67,24 +66,106 @@
     private static final String WIKIMEDIA_NEWS_ARTICLE = "e:Q17633526";
 
     private static final String WIKIMEDIA_NAVIGATIONAL_TEMPLATE = "e:Q11753321";
-    /**
-     *
-     * @param tokens the words spanned by the mention
-     * @param limit maximum number of results
-     * @return a query to retrieve candidate entities
-     */
-    public static TupleQuery generateCandidateQuery(RepositoryConnection conn, String tokens,
-        int limit, IRI aDescriptionIri)
+
+    private static String getExactMatchingQueryPart(String aString)
     {
+        return String.join("\n",
+            "    SELECT DISTINCT ?e2 ?description WHERE",
+            "    {",
+            "     VALUES ?labelpredicate {rdfs:label skos:altLabel}",
+            "      {",
+            "        ?e2 ?labelpredicate ?" + aString + " @en .",
+            "        OPTIONAL",
+            "        {",
+            "          ?e2 ?descriptionIri ?description.",
+            "          FILTER ( lang(?description) = \"en\" )",
+            "        }",
+            "      }",
+            "    }");
+    }
+
+    public static TupleQuery generateCandidateExactQuery(RepositoryConnection conn,
+        String aTypedString, String aMention, IRI aDescriptionIri)
+    {
+        // Matching user input exactly
+        String exactMatchingTypedString = getExactMatchingQueryPart("exactTyped");
+
+        // Usually users are lazy and don't capitalize words,
+        // but then the exact matching wouldn't find anything.
+        String exactMatchingTypedStringCapitalized = getExactMatchingQueryPart("exactTypedCapitalized");
+
+        // Do the same for the mention
+        String exactMatchingMention = getExactMatchingQueryPart("exactMention");
+
+        String exactMatchingMentionCapitalized = getExactMatchingQueryPart("exactMentionCapitalized");
+
         String query = String.join("\n",
             "DEFINE input:inference 'instances'",
             SPARQL_PREFIX,
             "SELECT DISTINCT ?e2 ?altLabel ?label ?description WHERE",
             "{",
             "  {",
+            exactMatchingTypedString,
```

Comment:
This looks susceptible to SPARQL injection.

Reply:
I will escape it.